src/Lean/Meta/Tactic/Rewrites.lean

```diff
@@ -0,0 +1,378 @@
+      withReducible (← mkFreshExprMVar r.result.eNew).mvarId!.applyRfl
+      -- We do not need to record the updated `MetavarContext` here.
+      pure true
+  catch _e =>
+    pure false
+
+/--
+Pretty print the result of the rewrite.
+-/
+def RewriteResult.ppResult (r : RewriteResult) : MetaM String :=
+  if let some pp := r.ppResult? then
+    return pp
+  else
+    return (← ppExpr r.result.eNew).pretty
+
+
+/-- Should we try discharging side conditions? If so, using `assumption`, or `solve_by_elim`? -/
+inductive SideConditions
+| none
+| assumption
+| solveByElim
+
+/-- Shortcut for calling `solveByElim`. -/
+def solveByElim (goals : List MVarId) (depth : Nat := 6) : MetaM PUnit := do
+  -- There is only a marginal decrease in performance for using the `symm` option for `solveByElim`.
+  -- (measured via `lake build && time lake env lean test/librarySearch.lean`).
+  let cfg : SolveByElimConfig := { maxDepth := depth, exfalso := false, symm := true }
+  let ⟨lemmas, ctx⟩ ← mkAssumptionSet false false [] [] #[]
+  let [] ← SolveByElim.solveByElim cfg lemmas ctx goals
+    | failure
+
+def rwLemma (ctx : MetavarContext) (goal : MVarId) (target : Expr) (side : SideConditions := .solveByElim) (lem : Expr ⊕ Name) (symm : Bool) (weight : Nat) : MetaM (Option RewriteResult) :=
+  withMCtx ctx do
+    let some expr ← (match lem with
+    | .inl hyp => pure (some hyp)
+    | .inr lem => some <$> mkConstWithFreshMVarLevels lem <|> pure none)
+      | return none
+    trace[Tactic.rewrites] m!"considering {if symm then "← " else ""}{expr}"
+    let some result ← some <$> goal.rewrite target expr symm <|> pure none
+      | return none
+    if result.mvarIds.isEmpty then
+      return some {
+        expr,
+        symm,
+        weight,
+        result,
+        ppResult? := none,
+        mctx := ← getMCtx
+      }
+    else
+      -- There are side conditions, which we try to discharge using local hypotheses.
+      match (← match side with
+        | .none => pure false
+        | .assumption => ((result.mvarIds.mapM fun m => m.assumption) >>= fun _ => pure true) <|> pure false
+        | .solveByElim => (solveByElim result.mvarIds >>= fun _ => pure true) <|> pure false) with
+      | false => return none
+      | true =>
+        -- If we succeed, we need to reconstruct the expression to report that we rewrote by.
+        let some expr := rewriteResultLemma result | return none
+        let expr ← instantiateMVars expr
+        let (expr, symm) := if expr.isAppOfArity ``Eq.symm 4 then
+            (expr.getArg! 3, true)
+          else
+            (expr, false)
+        return some {
+            expr, symm, weight, result,
+            ppResult? := none,
+            mctx := ← getMCtx
+          }
+
+/--
+Find keys which match the expression, or some subexpression.
+
+Note that repeated subexpressions will be visited each time they appear,
+making this operation potentially very expensive.
+It would be good to solve this problem!
```

Comment:
We could potentially address this now. I think the right approach would be to generalize/duplicate `Lean.Util.FindExpr` to allow accumulating a value rather than just returning a Bool.

Up to you (or I can take a look).

Reply:
How much of a problem has this been in practice?